source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+                // component_ids[id_index], inst_iter);
+              } else if (GetType(inst)->AsStruct()) {
+                assert(false &&
+                       "OpSpecConstantOp with struct type constants is not "
+                       "suppported");
+                // InsertConstantAt(
+                // GetType(inst)->AsStruct()->element_type(id_index),
+                // component_ids[id_index], inst_iter);
+              } else {
+                assert(false &&
+                       "New component constants should only be required when "
+                       "the result is vector, struct or array type constant");
+              }
+            }
+            break;
+        }
+        if (result) {
+          uint32_t result_id = GetNewId();
+          id_to_const_val_[result_id] = std::move(result);
+          inst_iter = InsertConstantAt(GetType(inst), result_id, inst_iter);
+          def_use_mgr_->ReplaceAllUsesWith(inst->result_id(), result_id);
+          def_use_mgr_->KillDef(inst->result_id());
+          inst->ToNop();
+          modified = true;
+        }
+        break;
+      }
+      default:
+        break;
+    }
+  } while (++inst_iter != module->types_values_end());
+  return modified;
+}
+
+std::unique_ptr<Value>
+FoldSpecConstantOpAndCompositePass::GetCompositeExtractResult(
+    ir::Instruction* inst) {
+  assert(inst->NumInOperands() - 1 == 2 &&
+         "OpSpecConstantOp CompositeExtract requires two non-type non-opcode "
+         "operands.");
+  assert(inst->GetInOperand(1).type == SPV_OPERAND_TYPE_ID &&
+         "The vector operand must have a SPV_OPERAND_TYPE_ID type");
+  assert(
+      inst->GetInOperand(2).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+      "The literanl operand must have a SPV_OPERAND_TYPE_LITERAL_INTEGER type");
+  auto iter_id = id_to_const_val_.find(inst->GetSingleWordInOperand(1));
+  if (iter_id != id_to_const_val_.end()) {
+    assert(iter_id->second->AsComposite() &&
+           "The vector operand must have a composite type constant value");
+    CompositeValue* composite_val = iter_id->second->AsComposite();
+    uint32_t literal = inst->GetSingleWordInOperand(2);
+    assert(literal < composite_val->nested.size() &&
+           "Literal index out of bound of the composite constant");
+    auto iter_result = id_to_const_val_.find(composite_val->nested[literal]);
+    if (iter_result == id_to_const_val_.end()) return nullptr;
+    Value* result_val = iter_result->second.get();
+    // Make copy in unique_ptr and return as result. There might be a way to
+    // avoid the copy?
+    if (result_val->AsScalar()) {
+      return std::unique_ptr<Value>(new ScalarValue(result_val->AsScalar()));
+    } else {
+      return std::unique_ptr<Value>(
+          new CompositeValue(result_val->AsComposite()));
+    }
+  } else {
+    // Does not find the composite operand in the id_to_const_val_ map, the
+    // value of the operand might not be fully resolved yet, so this
```

Comment:
All ids referenced should precede this instruction, thus should already be processed. ?